redis/redis.go

```diff
@@ -213,61 +227,77 @@ func (c *Client) publishMessageFromBuffer(s *bufferMessage) {
 		})
 
 		if mErr == nil {
-			c.client.Publish(c.ctx, SyncChannelName, binMsg)
+			c.client.Publish(ctx, SyncChannelName, binMsg)
 		}
 
-		c.client.Set(c.ctx,
+		c.client.Set(ctx,
 			prefixKey(s.Key),
 			binRes,
 			c.getTTL(origRes))
 	}
 }
 
-func (c *Client) processReceivedMessage(msg *redis.Message) {
+func (c *Client) processReceivedMessage(ctx context.Context, msg *redis.Message) {
+	select {
+	case <-ctx.Done():
+		return
+	default:
+	}
+
 	var rm redisMessage
 
-	err := json.Unmarshal([]byte(msg.Payload), &rm)
-	if err == nil {
-		// message was sent from a different blocky instance
-		if !bytes.Equal(rm.Client, c.id) {
-			switch rm.Type {
-			case messageTypeCache:
-				var cm *CacheMessage
-
-				cm, err = convertMessage(&rm, 0)
-				if err == nil {
-					c.CacheChannel <- cm
-				}
-			case messageTypeEnable:
-				err = c.processEnabledMessage(&rm)
+	if err := json.Unmarshal([]byte(msg.Payload), &rm); err != nil {
+		c.l.Error("Processing error: ", err)
+
+		return
+	}
+
+	// message was sent from a different blocky instance
+	if !bytes.Equal(rm.Client, c.id) {
+		switch rm.Type {
+		case messageTypeCache:
+			var cm *CacheMessage
+
+			cm, err := convertMessage(&rm, 0)
+			if err != nil {
+				c.l.Error("Processing CacheMessage error: ", err)
+
+				return
+			}
+
+			select {
+			case <-ctx.Done():
+				return
 			default:
-				c.l.Warn("Unknown message type: ", rm.Type)
+				c.CacheChannel <- cm
 			}
-		}
-	}
+		case messageTypeEnable:
+			var msg EnabledMessage
 
-	if err != nil {
-		c.l.Error("Processing error: ", err)
-	}
-}
+			if err := json.Unmarshal(rm.Message, &msg); err != nil {
+				c.l.Error("Processing EnabledMessage error: ", err)
 
-func (c *Client) processEnabledMessage(redisMsg *redisMessage) error {
-	var msg EnabledMessage
+				return
+			}
 
-	err := json.Unmarshal(redisMsg.Message, &msg)
-	if err == nil {
-		c.EnabledChannel <- &msg
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				c.EnabledChannel <- &msg
```

Comment:
Same as above, the send could be blocking, so let's make sure it cancelable:
```suggestion
			case c.EnabledChannel <- &msg:
```